src/make_knowledge_packs.jl

```diff
@@ -0,0 +1,222 @@
+    # sources_pattern = r"^(.*)-sources-(\d+)\.jls$"
+
+    # Group files by hostname and chunk size
+    for file in entries
+        match_chunks = match(chunks_pattern, file)
+        match_sources = match(sources_pattern, file)
+
+        if match_chunks !== nothing
+            hostname = match_chunks.captures[1]
+            chunk_size = parse(Int, match_chunks.captures[2])
+            if !haskey(hostname_files, hostname)
+                hostname_files[hostname] = Dict{Int, Dict{String, String}}()
+            end
+            if !haskey(hostname_files[hostname], chunk_size)
+                hostname_files[hostname][chunk_size] = Dict{String, String}()
+            end
+            hostname_files[hostname][chunk_size]["chunks"] = joinpath(
+                knowledge_pack_path, file)
+        elseif match_sources !== nothing
+            hostname = match_sources.captures[1]
+            chunk_size = parse(Int, match_sources.captures[2])
+            if !haskey(hostname_files, hostname)
+                hostname_files[hostname] = Dict{Int, Dict{String, String}}()
+            end
+            if !haskey(hostname_files[hostname], chunk_size)
+                hostname_files[hostname][chunk_size] = Dict{String, String}()
+            end
+            hostname_files[hostname][chunk_size]["sources"] = joinpath(
+                knowledge_pack_path, file)
+        end
+    end
+    # Process each pair of files
+    for (hostname, chunk_files) in hostname_files
+        for (chunk_size, files) in chunk_files
+            if haskey(files, "chunks") && haskey(files, "sources")
+                chunks_file = files["chunks"]
+                sources_file = files["sources"]
+                chunks = deserialize(chunks_file)
+                sources = deserialize(sources_file)
+                cost_tracker = Threads.Atomic{Float64}(0.0)
+                full_embeddings = RT.get_embeddings(
+                    embedder, chunks; model, verbose = false, cost_tracker)
+                @info "Created embeddings for $hostname. Cost: \$$(round(cost_tracker[], digits=3))"
+                fn_output = joinpath(knowledge_pack_path, "packs",
+                    "$hostname-textembedding3large-0-Float32__v1.0.tar.gz")
+                fn_temp = joinpath(knowledge_pack_path, "packs",
+                    "$hostname-textembedding3large-0-Float32__v1.0.hdf5")
+                h5open(fn_temp, "w") do file
+                    file["chunks"] = chunks
+                    file["sources"] = sources
+                    file["embeddings"] = full_embeddings[1:embedding_size, :] |>
+                                         l2_norm_columns |> x -> map(>(0), x)
+                    file["type"] = "ChunkIndex"
+                    # file["metadata"] = "$hostname ecosystem docstrings, chunk size $chunk_size, downloaded on 20240330, contains: Makie.jl, AlgebraOfGraphics.jl, GeoMakie.jl, GraphMakie.jl, MakieThemes.jl, TopoPlots.jl, Tyler.jl"
```

Comment:
Could you please add some metadata, at least the date stamp when it was processed... `Dates.today()` ideally, also when it was scraped, eg, `Dict(:scraped_dt=>Date(...), :embedded_dt=>Dates.today(), :packages=>[url1,url2]`)`? You can probably extract the package names from the URLs in sources by keeping only the main package name?

Reply:
I added some metadata: date, package_urls,, chunk size, model, embedding_size custom_metadata string (if the user wants to add a custom string like some ecosystem).

I can extract the package name but there's an issue of consistency. some doc websites have the package name with ".jl" in the url, some have it in the side bar. These are identifiable. But some doc websites have the package name in url without ".jl" so it becomes difficult to identify it. I can put the hostname of url or "unknown". What do you suggest?

Reply:
Take the URLs you have now and write some simple rules /regex to do 80% job. it's better to attach some package names then none (easier than loading the whole pack and parsing the sources).

If package names are too hard, strip the URL junk and return base URL with first path segment?